src/err/err_state.rs

```diff
@@ -16,13 +16,19 @@ pub(crate) struct PyErrStateNormalized {
 
 impl PyErrStateNormalized {
     #[cfg(not(Py_3_12))]
-    pub(crate) fn ptype<'py>(&'py self, py: Python<'py>) -> &'py PyType {
-        self.ptype.as_ref(py)
+    pub(crate) fn ptype<'py>(&'py self, py: Python<'py>) -> Bound<'py, PyType> {
+        self.ptype.bind(py).clone()
     }
 
     #[cfg(Py_3_12)]
-    pub(crate) fn ptype<'py>(&'py self, py: Python<'py>) -> &'py PyType {
-        self.pvalue.as_ref(py).get_type()
+    pub(crate) fn ptype<'py>(&'py self, py: Python<'py>) -> Bound<'py, PyType> {
+        use crate::ffi_ptr_ext::FfiPtrExt;
+        use crate::types::any::PyAnyMethods;
+        unsafe {
+            ffi::PyExceptionInstance_Class(self.pvalue.as_ptr())
+                .assume_owned(py)
+                .downcast_into_unchecked()
+        }
```

Comment:
Resourceful to go through the C API but I think that this introduces a refcounting error from `PyExceptionInstance_Class` actually returning a borrowed pointer (because it's just a wrapper around `Py_TYPE` which also just borrows the pointer).

Probably it's better to keep using our safe wrappers with something like

```
        self.pvalue.bind(py).get_type()
```

As you probably noticed, `PyAnyMethods::get_type` doesn't yet return `Bound<PyType>`.

Maybe for now just add some compatibility and I'll fixup in #3705 later:

```suggestion
        self.pvalue.bind(py).get_type().as_borrowed().to_owned()
```

Reply:
Sure!

I was trying to follow the example of how `fn ptraceback` is implemented below for `Py_3_12`. I'm curious what the difference here is?

Reply:
I think it's probably a historical difference that might have arisen from the fact that reading an object's type uses these borrowed pointers for performance (even though it's unsafe to borrow as `__class__` can be assigned to, even if that's not a great idea). `PyExceptionInstance_Class` isn't actually documented on https://docs.python.org/3/c-api/exceptions.html. I suspect that these days it's always just assumed that `Py_TYPE` (or our equivalent in PyO3) is good enough.

In contrast, [`PyException_GetTraceback`](https://docs.python.org/3/c-api/exceptions.html#c.PyException_GetTraceback) _is_ documented and states it returns a new (i.e. owned) reference.